### ISSUE TITLE
FastAPI endpoint that contains IO bound and CPU bound tasks. better with or without async??

### ISSUE THREAD

**Issue by oriolcmp** [tiangolo/fastapi#3725]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
I don't really have code to share
```

### Description

Hello,

I'm trying to move from Flask to FastAPI because it seems better prepared to work concurrently. But the API that I'm trying to build have several endpoints with the same pattern. 
When they get a request, they connect to a DynamoDB and do a query. After getting that information they do some CPU intensive task (a few seconds) and then it saves the new data in the DB and returns that data.
I'm not sure what the best way is in that case. Could the **async def** syntax help me in this case? What is the current best approach to handle this issue?
Thanks,

### Operating System

Linux

### FastAPI Version

0.68.0

### Python Version

3.8.8

**Comment by frankie567** [comment]
Here is some technical details about FastAPI and how it's handling async/sync operations (see: https://fastapi.tiangolo.com/async/?h=technical#very-technical-details):

* If you define an endpoint with `async`, it'll run in the event loop; even if there are blocking calls inside (like a blocking call to your database). It means that it can block the server thread and prevent it to answer other requests meanwhile. That's why it's preferable to use async-compatible libraries for I/O operations.
* If you define an endpoint without `async`, as a standard function, FastAPI will **automatically** send it in a separate thread to avoid the blocking of the event loop. In a sense, it "mimics" an asynchronous behavior.
  * Having to send the task to a separate thread implies a small overhead; so unless you are really doing heavy blocking I/O, it's not worth it.

Concerning your case, here are my advices:

 * Use an async-compatible client for DynamoDB (https://aioboto3.readthedocs.io/en/latest/index.html). This way, you can define every endpoints as `async` and benefit from async I/O goodness.
 * You say that CPU operations take a few seconds. Async I/O won't help you in this context. If you can live with API endpoints that take a few seconds to answer, well, that should do it.
   * If you really need your API to answer faster, you should consider use a worker process. Basically, it's a separate process which role is to handle slow background operations. Your API will just send messages in a task queue and respond immediately.
   * On the other side, the worker will handle those tasks as they come. The only pitfall is that your API can't return the result synchronously. You just know that the operation will be handled at some point. Thus, you'll have to implement separate endpoints to check for the completion and results of those tasks.
   * Libraries like [Celery](https://docs.celeryproject.org/en/stable/) are designed to do this.

**Comment by oriolcmp** [comment]
Thank you very much for this deep explanation. I helps me to understand the whole process.
Actually, I have tried your suggestion to add the `aioboto3` library so my endpoint can be a real async endpoint. I even added ` await run_in_threadpool` for the method that runs the CPU expensive task to make it fully async. But I was surprised to see its performance. Because I did an stress test by sending 150 requests to the endpoint in 5 seconds. This test I also run it for the code versions where there was just the difference of the `async def` syntax in the endpoint definition. Now, with the new endpoint fully async I also run the test and I see these results:

~4min when the endoint is defined as `def` or just `async def` but using `boto3`
~4min 30s when the endpoint uses the full async capacity with `aioboto3` and `run_in_threadpool`

So it seems is better to leave it without async syntax or even just the async syntax without any await works better. 
I just wanted to comment it here, maybe it is normal but I didn't expect that.

**Comment by frankie567** [comment]
Have you tried with `aioboto3` but keep your computation in the main process (without `run_in_threadpool`)? I would be curious to see the results. As mentioned, sending a task to a threadpool adds some overhead, so it may not worth it in your case.

Once again, if you really need high scalability and reliability, you'll have to go for the worker model; but it obviously makes the architecture of your system more complex.

**Comment by oriolcmp** [comment]
Yes, I first tested it with `run_in_threadpool` and then without  and the runtime of the test was very similar (few seconds difference). 
I'm not worried if the endpoint takes a few seconds to answer but I'm trying to take as much as possible advantage of the concurrency so the server can answer the most possible requests. That's why I was trying FastAPI to compare with Flask, but with this special case I see very similar performance in both of them.

**Comment by raphaelauv** [comment]
@oriolcmp look this -> https://github.com/encode/starlette/issues/1094

You can very probably close the issue , thank you

**Comment by anguspmitchell** [comment]
@frankie567 I'm in a similar situation. First I'm doing non-async I/O (postgres query), then I'm doing some processing on each result. The processing I s just a series of maps, so if there were a way to parallelize it, it would speed up my responses considerably.

I have a very poor understanding of processes/threads, but is there a way to create multiple processes from inside a `def` endpoint? (Or would it be created once, by the app, and not on each request?)

I saw a SO response on this, but I couldn't get this working -- my uvicorn server just kept rebooting. https://stackoverflow.com/a/63171013

**Comment by frankie567** [comment]
@anguspmitchell Generally speaking, I would avoid spawning processes/threads inside a web request. A request should consist of quick operations (a read/write I/O, JSON serialization, template rendering...).

> First I'm doing non-async I/O (postgres query), then I'm doing some processing on each result. The processing is just a series of maps, so if there were a way to parallelize it, it would speed up my responses considerably.

If that's the case, it probably means that you're trying to handle too much data in one request. My advice would be to implement a pagination mechanism so you can lookup for only a limited number of rows that can be processed in a reasonable time.

If you *really* need to handle all the data at once, a better solution would be to have a separate worker as I described here: https://github.com/tiangolo/fastapi/issues/3725#issuecomment-902629033
